daemon/prompter.c

```diff
@@ -999,6 +1009,56 @@ prompter_cancel_invocation(prompter_t *self)
     }
 }
 
+static void
+prompter_watch_name(prompter_t *self, GDBusConnection *connection, const gchar *name)
+{
+    if( !g_hash_table_contains(self->prm_watchers, name) ) {
+        guint id = g_bus_watch_name_on_connection(
+                connection, name, G_BUS_NAME_OWNER_FLAGS_NONE, NULL,
+                prompter_handle_name_lost, self, NULL);
+        g_hash_table_insert(self->prm_watchers, g_strdup(name), GUINT_TO_POINTER(id));
+        log_debug("watching for '%s' to leave bus", name);
+    }
+}
+
+static void
+prompter_handle_name_lost(GDBusConnection *connection, const gchar *name, gpointer user_data)
+{
+    (void)connection; // unused
+    prompter_t *self    = user_data;
+    guint       watcher = GPOINTER_TO_UINT(g_hash_table_lookup(self->prm_watchers, name));
+    log_debug("'%s' left bus", name);
+
+    /* First check the current invocation */
+    if( self->prm_invocation &&
+        !g_strcmp0(g_dbus_method_invocation_get_sender(self->prm_invocation), name) ) {
+        GDBusMethodInvocation *invocation = self->prm_invocation;
+        log_debug("-> canceling %p", invocation);
+        prompter_cancel_invocation(self);
+        prompter_return_error(invocation, G_DBUS_ERROR_AUTH_FAILED,
+                              SERVICE_MESSAGE_DISCONNECTED);
+        prompter_eval_state_later(self);
+    }
+
+    /* Then check the rest of the invocations */
+    GList *iter = prompter_iter(self);
+    while( iter ) {
+        GDBusMethodInvocation *invocation = iter->data;
+        if( !g_strcmp0(g_dbus_method_invocation_get_sender(invocation), name) ) {
+            log_debug("-> skipping %p", invocation);
+            prompter_return_error(invocation, G_DBUS_ERROR_AUTH_FAILED,
+                                  SERVICE_MESSAGE_DISCONNECTED);
+            iter = prompter_drop(self, iter);
+        } else {
+            iter = iter->next;
+        }
+    }
+
+    if( watcher )
+        g_bus_unwatch_name(watcher);
+    g_hash_table_remove(self->prm_watchers, name);
```

Comment:
Is there some need to delay watcher cleanup until here?
I'd consider adding prompter_unwatch_name() counterpart for prompter_watch_name(), hide the id/hash table details in there, and then call it at the start of this function.

Reply:
No reason. Changed it like you suggested.